Future-proof against breaking Kafka 3.5 changes

The upcoming Kafka 3.5 release comes with a change that will break Cruise Control. 

Most notable change is the package name change of:
 
`org.apache.kafka.server.log.internals` -> `org.apache.kafka.storage.internals.log` [1]

This change will cause issues with the metric reporter when run it with Kafka 3.5 brokers:
```
2023-05-23 17:17:15,315 ERROR Exiting Kafka due to fatal exception during startup. (kafka.Kafka$) [main]
java.lang.NoClassDefFoundError: kafka/log/LogConfig
        at com.linkedin.kafka.cruisecontrol.metricsreporter.CruiseControlMetricsReporter.createMetricsTopicFromReporterConfig(CruiseControlMetricsReporter.java:287)
        at com.linkedin.kafka.cruisecontrol.metricsreporter.CruiseControlMetricsReporter.configure(CruiseControlMetricsReporter.java:196)
        at org.apache.kafka.common.config.AbstractConfig.getConfiguredInstance(AbstractConfig.java:401)
        at org.apache.kafka.common.config.AbstractConfig.getConfiguredInstances(AbstractConfig.java:482)
        at kafka.server.DynamicMetricReporterState.createReporters(DynamicBrokerConfig.scala:871)
        at kafka.server.DynamicMetricReporterState.<init>(DynamicBrokerConfig.scala:857)
        at kafka.server.DynamicMetricsReporters.<init>(DynamicBrokerConfig.scala:797)
        at kafka.server.DynamicBrokerConfig.addReconfigurables(DynamicBrokerConfig.scala:265)
        at kafka.server.KafkaServer.startup(KafkaServer.scala:548)
        at kafka.Kafka$.main(Kafka.scala:113)
        at kafka.Kafka.main(Kafka.scala)
Caused by: java.lang.ClassNotFoundException: kafka.log.LogConfig
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
        at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:520)
        ... 11 more
```
Since the metric reporter cannot find the the LogConfig class, because the location of the LogConfig class changes with Kafka 3.5.


[1] https://github.com/apache/kafka/pull/13213/files#diff-9d7d92cafcbea44968abada4825979d4eefa9d07bf49c444ce11c0c24b327127

--- a/build.gradle
+++ b/build.gradle
@@ -278,6 +278,8 @@ project(':cruise-control') {
     implementation "io.netty:netty-transport-native-epoll:${nettyVersion}"
     api "org.apache.kafka:kafka_$scalaBinaryVersion:$kafkaVersion"
     api "org.apache.kafka:kafka-clients:$kafkaVersion"
+    // Add following dependency when upgrading to Kafka 3.5
+    //api "org.apache.kafka:kafka-storage:$kafkaVersion"
     implementation "org.scala-lang:scala-library:$scalaVersion"
     implementation 'org.apache.commons:commons-math3:3.6.1'
     implementation 'org.apache.httpcomponents:httpclient:4.5.13'

--- a/cruise-control-metrics-reporter/src/main/java/com/linkedin/kafka/cruisecontrol/metricsreporter/CruiseControlMetricsReporter.java
+++ b/cruise-control-metrics-reporter/src/main/java/com/linkedin/kafka/cruisecontrol/metricsreporter/CruiseControlMetricsReporter.java
@@ -25,7 +25,6 @@ import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.TimeoutException;
-import kafka.log.LogConfig;
 import kafka.server.KafkaConfig;
 import org.apache.kafka.clients.ClientUtils;
 import org.apache.kafka.clients.CommonClientConfigs;
@@ -284,9 +283,9 @@ public class CruiseControlMetricsReporter implements MetricsReporter, Runnable {
     NewTopic newTopic = new NewTopic(cruiseControlMetricsTopic, cruiseControlMetricsTopicNumPartition, cruiseControlMetricsTopicReplicaFactor);
 
     Map<String, String> config = new HashMap<>();
-    config.put(LogConfig.RetentionMsProp(),
+    config.put(TopicConfig.RETENTION_MS_CONFIG,
                Long.toString(reporterConfig.getLong(CruiseControlMetricsReporterConfig.CRUISE_CONTROL_METRICS_TOPIC_RETENTION_MS_CONFIG)));
-    config.put(LogConfig.CleanupPolicyProp(), CRUISE_CONTROL_METRICS_TOPIC_CLEAN_UP_POLICY);
+    config.put(TopicConfig.CLEANUP_POLICY_CONFIG, CRUISE_CONTROL_METRICS_TOPIC_CLEAN_UP_POLICY);
     if (cruiseControlMetricsTopicMinInsyncReplicas > 0) {
       // If the user has set the minISR for the metrics topic we need to check that the replication factor is set to a level that allows the
       // minISR to be met.
@@ -358,8 +357,8 @@ public class CruiseControlMetricsReporter implements MetricsReporter, Runnable {
       Config topicConfig = describeConfigsResult.values().get(topicResource).get(CLIENT_REQUEST_TIMEOUT_MS, TimeUnit.MILLISECONDS);
       Set<AlterConfigOp> alterConfigOps = new HashSet<>();
       Map<String, String> configsToSet = new HashMap<>();
-      configsToSet.put(LogConfig.RetentionMsProp(), _metricsTopic.configs().get(LogConfig.RetentionMsProp()));
-      configsToSet.put(LogConfig.CleanupPolicyProp(), _metricsTopic.configs().get(LogConfig.CleanupPolicyProp()));
+      configsToSet.put(TopicConfig.RETENTION_MS_CONFIG, _metricsTopic.configs().get(TopicConfig.RETENTION_MS_CONFIG));
+      configsToSet.put(TopicConfig.CLEANUP_POLICY_CONFIG, _metricsTopic.configs().get(TopicConfig.CLEANUP_POLICY_CONFIG));
       maybeUpdateConfig(alterConfigOps, configsToSet, topicConfig);
       if (!alterConfigOps.isEmpty()) {
         AlterConfigsResult alterConfigsResult = _adminClient.incrementalAlterConfigs(Collections.singletonMap(topicResource, alterConfigOps));

--- a/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/KafkaCruiseControlUtils.java
+++ b/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/KafkaCruiseControlUtils.java
@@ -43,6 +43,7 @@ import org.apache.kafka.common.config.ConfigException;
 import org.apache.kafka.common.config.ConfigResource;
 import org.apache.kafka.common.config.SaslConfigs;
 import org.apache.kafka.common.config.SslConfigs;
+import org.apache.kafka.common.config.TopicConfig;
 import org.apache.kafka.common.errors.ReassignmentInProgressException;
 import org.apache.kafka.common.errors.TopicExistsException;
 import org.apache.kafka.common.message.MetadataResponseData;
@@ -80,8 +81,6 @@ import scala.Option;
 
 import static com.linkedin.kafka.cruisecontrol.config.constants.MonitorConfig.RECONNECT_BACKOFF_MS_CONFIG;
 import static com.linkedin.kafka.cruisecontrol.servlet.parameters.ParameterUtils.SKIP_HARD_GOAL_CHECK_PARAM;
-import static kafka.log.LogConfig.CleanupPolicyProp;
-import static kafka.log.LogConfig.RetentionMsProp;
 
 
 /**
@@ -243,8 +242,8 @@ public final class KafkaCruiseControlUtils {
 
     NewTopic newTopic = new NewTopic(topic, partitionCount, replicationFactor);
     Map<String, String> config = new HashMap<>();
-    config.put(RetentionMsProp(), Long.toString(retentionMs));
-    config.put(CleanupPolicyProp(), DEFAULT_CLEANUP_POLICY);
+    config.put(TopicConfig.RETENTION_MS_CONFIG, Long.toString(retentionMs));
+    config.put(TopicConfig.CLEANUP_POLICY_CONFIG, DEFAULT_CLEANUP_POLICY);
     newTopic.configs(config);
 
     return newTopic;

--- a/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ReplicationThrottleHelper.java
+++ b/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ReplicationThrottleHelper.java
@@ -424,40 +424,36 @@ class ReplicationThrottleHelper {
       // First we try to get the LogConfig class for Kafka 3.5+
       logConfigClass = Class.forName(LOG_CONFIG_IN_KAFKA_3_5_AND_LATER);
       LOG.info("Found class {} for Kafka 3.5 and newer.", LOG_CONFIG_IN_KAFKA_3_5_AND_LATER);
+      try {
+        Field field = logConfigClass.getDeclaredField(config.toString());
+        return field.get(null).toString();
+      } catch (NoSuchFieldException | IllegalAccessException e) {
+        LOG.info("Field {} not found. We are probably on Kafka 3.4 or older.", config);
+      }
     } catch (ClassNotFoundException e) {
       LOG.info("Class {} not found. We are probably on Kafka 3.4 or older.", LOG_CONFIG_IN_KAFKA_3_5_AND_LATER);
+    }
 
-      // We did not find the LogConfig class from Kafka 3.5+.
-      // So we are probably on older Kafka version => we will try the older class for Kafka 3.4-.
+    // We did not find the LogConfig class or field from Kafka 3.5+.
+    // So we are probably on older Kafka version => we will try the older class for Kafka 3.4-.
+    try {
+      logConfigClass = Class.forName(LOG_CONFIG_IN_KAFKA_3_4_AND_EARLIER);
+      LOG.info("Found class {} for Kafka 3.4 and earlier.", LOG_CONFIG_IN_KAFKA_3_4_AND_EARLIER);
       try {
-        logConfigClass = Class.forName(LOG_CONFIG_IN_KAFKA_3_4_AND_EARLIER);
-        LOG.info("Found class {} for Kafka 3.4 and earlier.", LOG_CONFIG_IN_KAFKA_3_4_AND_EARLIER);
-      } catch (ClassNotFoundException ee) {
-          // No class was found for any Kafka version => we should fail
-          throw new RuntimeException("Failed to find LogConfig class", ee);
+        String nameOfMethod = "";
+        if (config == LogConfig.LEADER_REPLICATION_THROTTLED_REPLICAS_CONFIG) {
+          nameOfMethod = "LeaderReplicationThrottledReplicasProp";
+        } else if (config == LogConfig.FOLLOWER_REPLICATION_THROTTLED_REPLICAS_CONFIG) {
+          nameOfMethod = "FollowerReplicationThrottledReplicasProp";
         }
-    }
-
-    try {
-      Field field = logConfigClass.getDeclaredField(config.toString());
-      return field.get(null).toString();
-    } catch (NoSuchFieldException | IllegalAccessException e) {
-      LOG.info("Field {} not found. We are probably on Kafka 3.4 or older.", config);
-
-      // We did not find the LogConfig class field from Kafka 3.5+ LogConfig class.
-      // So we are probably on older Kafka version => we will try the older class for Kafka 3.4-.
-    }
-    try {
-      String nameOfMethod = "";
-      if (config == LogConfig.LEADER_REPLICATION_THROTTLED_REPLICAS_CONFIG) {
-        nameOfMethod = "LeaderReplicationThrottledReplicasProp";
-      } else if (config == LogConfig.FOLLOWER_REPLICATION_THROTTLED_REPLICAS_CONFIG) {
-        nameOfMethod = "FollowerReplicationThrottledReplicasProp";
+        Method method = logConfigClass.getMethod(nameOfMethod);
+        return method.invoke(null).toString();
+      } catch (InvocationTargetException | NoSuchMethodException | IllegalAccessException e) {
+        throw new RuntimeException("Failed to get configuration", e);
       }
-      Method method = logConfigClass.getMethod(nameOfMethod);
-      return method.invoke(null).toString();
-    } catch (InvocationTargetException | NoSuchMethodException | IllegalAccessException e) {
-      throw new RuntimeException("Failed to get configuration", e);
+    } catch (ClassNotFoundException e) {
+      // No class was found for any Kafka version => we should fail
+      throw new RuntimeException("Failed to find LogConfig class", e);
     }
   }
 }

--- a/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ReplicationThrottleHelper.java
+++ b/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ReplicationThrottleHelper.java
@@ -423,37 +423,30 @@ class ReplicationThrottleHelper {
     try {
       // First we try to get the LogConfig class for Kafka 3.5+
       logConfigClass = Class.forName(LOG_CONFIG_IN_KAFKA_3_5_AND_LATER);
-      LOG.info("Found class {} for Kafka 3.5 and newer.", LOG_CONFIG_IN_KAFKA_3_5_AND_LATER);
-      try {
-        Field field = logConfigClass.getDeclaredField(config.toString());
-        return field.get(null).toString();
-      } catch (NoSuchFieldException | IllegalAccessException e) {
-        LOG.info("Field {} not found. We are probably on Kafka 3.4 or older.", config);
-      }
-    } catch (ClassNotFoundException e) {
-      LOG.info("Class {} not found. We are probably on Kafka 3.4 or older.", LOG_CONFIG_IN_KAFKA_3_5_AND_LATER);
+
+      Field field = logConfigClass.getDeclaredField(config.toString());
+      return field.get(null).toString();
+    } catch (ClassNotFoundException | NoSuchFieldException | IllegalAccessException e) {
+      LOG.info("Failed to read config {} from LogConfig class since we are probably on kafka 3.4 or older: {}", config, e);
     }
 
     // We did not find the LogConfig class or field from Kafka 3.5+.
     // So we are probably on older Kafka version => we will try the older class for Kafka 3.4-.
     try {
       logConfigClass = Class.forName(LOG_CONFIG_IN_KAFKA_3_4_AND_EARLIER);
-      LOG.info("Found class {} for Kafka 3.4 and earlier.", LOG_CONFIG_IN_KAFKA_3_4_AND_EARLIER);
-      try {
-        String nameOfMethod = "";
-        if (config == LogConfig.LEADER_REPLICATION_THROTTLED_REPLICAS_CONFIG) {
-          nameOfMethod = "LeaderReplicationThrottledReplicasProp";
-        } else if (config == LogConfig.FOLLOWER_REPLICATION_THROTTLED_REPLICAS_CONFIG) {
-          nameOfMethod = "FollowerReplicationThrottledReplicasProp";
-        }
-        Method method = logConfigClass.getMethod(nameOfMethod);
-        return method.invoke(null).toString();
-      } catch (InvocationTargetException | NoSuchMethodException | IllegalAccessException e) {
-        throw new RuntimeException("Failed to get configuration", e);
+
+      String nameOfMethod = "";
+      if (config == LogConfig.LEADER_REPLICATION_THROTTLED_REPLICAS_CONFIG) {
+        nameOfMethod = "LeaderReplicationThrottledReplicasProp";
+      } else if (config == LogConfig.FOLLOWER_REPLICATION_THROTTLED_REPLICAS_CONFIG) {
+        nameOfMethod = "FollowerReplicationThrottledReplicasProp";
       }
-    } catch (ClassNotFoundException e) {
-      // No class was found for any Kafka version => we should fail
-      throw new RuntimeException("Failed to find LogConfig class", e);
+
+      Method method = logConfigClass.getMethod(nameOfMethod);
+      return method.invoke(null).toString();
+      } catch (ClassNotFoundException | InvocationTargetException | NoSuchMethodException | IllegalAccessException e) {
+      // No class or method was found for any Kafka version => we should fail
+      throw new RuntimeException("Failed to read config " + config + " from LogConfig class:", e);
+      }
     }
-  }
 }

--- a/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ReplicationThrottleHelper.java
+++ b/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ReplicationThrottleHelper.java
@@ -13,6 +13,9 @@ import org.apache.kafka.clients.admin.ConfigEntry;
 import org.apache.kafka.common.config.ConfigResource;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collection;
@@ -38,10 +41,12 @@ class ReplicationThrottleHelper {
   static final String WILDCARD_ASTERISK = "*";
   static final String LEADER_THROTTLED_RATE = "leader.replication.throttled.rate";
   static final String FOLLOWER_THROTTLED_RATE = "follower.replication.throttled.rate";
-  // TODO: Update to LogConfig.LEADER_REPLICATION_THROTTLED_REPLICAS_CONFIG for upgrade to Kafka 3.5
-  static final String LEADER_THROTTLED_REPLICAS = "leader.replication.throttled.replicas";
-  // TODO: Update to LogConfig.FOLLOWER_REPLICATION_THROTTLED_REPLICAS_CONFIG for upgrade to Kafka 3.5
-  static final String FOLLOWER_THROTTLED_REPLICAS = "follower.replication.throttled.replicas";
+  // LogConfig class in Kafka 3.5+
+  private static final String LOG_CONFIG_IN_KAFKA_3_5_AND_LATER = "org.apache.kafka.storage.internals.log.LogConfig";
+  // LogConfig class in Kafka 3.4-
+  private static final String LOG_CONFIG_IN_KAFKA_3_4_AND_EARLIER = "kafka.log.LogConfig";
+  static final String LEADER_THROTTLED_REPLICAS = getLogConfig(LogConfig.LEADER_REPLICATION_THROTTLED_REPLICAS_CONFIG);
+  static final String FOLLOWER_THROTTLED_REPLICAS = getLogConfig(LogConfig.FOLLOWER_REPLICATION_THROTTLED_REPLICAS_CONFIG);
   public static final long CLIENT_REQUEST_TIMEOUT_MS = TimeUnit.SECONDS.toMillis(30);
   static final int RETRIES = 30;
 
@@ -390,5 +395,69 @@ class ReplicationThrottleHelper {
       }
     }
     return true;
+  }
+
+  private enum LogConfig {
+    LEADER_REPLICATION_THROTTLED_REPLICAS_CONFIG,
+    FOLLOWER_REPLICATION_THROTTLED_REPLICAS_CONFIG,
+  }
+
+  /**
+   * Starting with Kafka 3.5.0, the location of the "LogConfig" class is "org.apache.kafka.storage.internals.log.LogConfig"
+   * and provides new constant fields in place of some methods of the original class.
+   *
+   *   - LogConfig class in Kafka 3.5+: org.apache.kafka.storage.internals.log.LogConfig
+   *   - LogConfig class in Kafka 3.4-: kafka.log.LogConfig
+   **
+   * The older LogConfig class does not work with the newer versions of Kafka. Therefore, if the new class exists, we use it and if
+   * it doesn't exist we will fall back on the older one.
+   *
+   * Once CC supports only 3.5.0 and newer, we can clean this up and use the LogConfig class from
+   * `org.apache.kafka.storage.internals.log.LogConfig`all the time.
+   * @param config LogConfig config name
+   * @return LogConfig config name in the format of a Kafka configuration property
+   */
+  private static String getLogConfig(LogConfig config) {
+    Class<?> logConfigClass;
+
+    try {
+      // First we try to get the LogConfig class for Kafka 3.5+
+      logConfigClass = Class.forName(LOG_CONFIG_IN_KAFKA_3_5_AND_LATER);
+      LOG.info("Found class {} for Kafka 3.5 and newer.", LOG_CONFIG_IN_KAFKA_3_5_AND_LATER);
+    } catch (ClassNotFoundException e) {
+      LOG.info("Class {} not found. We are probably on Kafka 3.4 or older.", LOG_CONFIG_IN_KAFKA_3_5_AND_LATER);
+
+      // We did not find the LogConfig class from Kafka 3.5+.
+      // So we are probably on older Kafka version => we will try the older class for Kafka 3.4-.
+      try {
+        logConfigClass = Class.forName(LOG_CONFIG_IN_KAFKA_3_4_AND_EARLIER);
+        LOG.info("Found class {} for Kafka 3.4 and earlier.", LOG_CONFIG_IN_KAFKA_3_4_AND_EARLIER);
+      } catch (ClassNotFoundException ee) {
+          // No class was found for any Kafka version => we should fail
+          throw new RuntimeException("Failed to find LogConfig class", ee);
+        }
+    }
+
+    try {
+      Field field = logConfigClass.getDeclaredField(config.toString());
+      return field.get(null).toString();
+    } catch (NoSuchFieldException | IllegalAccessException e) {
+      LOG.info("Field {} not found. We are probably on Kafka 3.4 or older.", config);
+
+      // We did not find the LogConfig class field from Kafka 3.5+ LogConfig class.
+      // So we are probably on older Kafka version => we will try the older class for Kafka 3.4-.
+    }
+    try {
+      String nameOfMethod = "";
+      if (config == LogConfig.LEADER_REPLICATION_THROTTLED_REPLICAS_CONFIG) {
+        nameOfMethod = "LeaderReplicationThrottledReplicasProp";
+      } else if (config == LogConfig.FOLLOWER_REPLICATION_THROTTLED_REPLICAS_CONFIG) {
+        nameOfMethod = "FollowerReplicationThrottledReplicasProp";
+      }
+      Method method = logConfigClass.getMethod(nameOfMethod);
+      return method.invoke(null).toString();
+    } catch (InvocationTargetException | NoSuchMethodException | IllegalAccessException e) {
+      throw new RuntimeException("Failed to get configuration", e);
+    }
   }
 }

--- a/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ReplicationThrottleHelper.java
+++ b/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ReplicationThrottleHelper.java
@@ -6,7 +6,6 @@ package com.linkedin.kafka.cruisecontrol.executor;
 
 import com.linkedin.kafka.cruisecontrol.metricsreporter.CruiseControlMetricsUtils;
 import com.linkedin.kafka.cruisecontrol.model.ReplicaPlacementInfo;
-import kafka.log.LogConfig;
 import org.apache.kafka.clients.admin.AdminClient;
 import org.apache.kafka.clients.admin.AlterConfigOp;
 import org.apache.kafka.clients.admin.Config;
@@ -39,8 +38,10 @@ class ReplicationThrottleHelper {
   static final String WILDCARD_ASTERISK = "*";
   static final String LEADER_THROTTLED_RATE = "leader.replication.throttled.rate";
   static final String FOLLOWER_THROTTLED_RATE = "follower.replication.throttled.rate";
-  static final String LEADER_THROTTLED_REPLICAS = LogConfig.LeaderReplicationThrottledReplicasProp();
-  static final String FOLLOWER_THROTTLED_REPLICAS = LogConfig.FollowerReplicationThrottledReplicasProp();
+  // TODO: Update to LogConfig.LEADER_REPLICATION_THROTTLED_REPLICAS_CONFIG for upgrade to Kafka 3.5
+  static final String LEADER_THROTTLED_REPLICAS = "leader.replication.throttled.replicas";
+  // TODO: Update to LogConfig.FOLLOWER_REPLICATION_THROTTLED_REPLICAS_CONFIG for upgrade to Kafka 3.5
+  static final String FOLLOWER_THROTTLED_REPLICAS = "follower.replication.throttled.replicas";
   public static final long CLIENT_REQUEST_TIMEOUT_MS = TimeUnit.SECONDS.toMillis(30);
   static final int RETRIES = 30;
 

